[CALCITE-3245] CompileException in Janino when a query contains a division between a Double and a BigDecimal

DonnyZone opened the pull request:
This PR is proposed to handle the exception caused by Java `BigDecimal` type in [CALCITE-3245](https://issues.apache.org/jira/browse/CALCITE-3245).

(1) Implement `getPrecision` and `getScale` in JavaType. 
    As BigDecimal will trigger the computation on precision and scale (`deriveDecimalPlusType` , `deriveDecimalMultiplyType` in `RelDataTypeSystem`), but the default `PRECISION_NOT_SPECIFIED` and `SCALE_NOT_SPECIFIED` value will cause the assertions.

(2) Specially handle cast on `BigDecimal.class`. 
    Current now, there are more than one implementations for data type conversion, such as **`Types.castIfNecessary`**, **`RexToLixTranslator.convert`**, **`RexToLixTranslator.translateCast, and etc`**. We need to unify them later.

danny0405 commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:
```diff
@@ -2278,9 +2278,12 @@ public Expression implement(
+      }
+      return Types.castIfNecessary(returnType, result);
     }
 
```
Can we move this `BigDecimal` cast logic into `Types.castIfNecessary` ?

DonnyZone replied:
@danny0405 To the best of my knowledge, data type conversion are implemented in many places (`Types.castIfNecessary`, `RexToLixTranslator.convert`, `RexToLixTranslator.translateCast`, etc.) with some overlaps. We can move this logic into `Types.castIfNecessary`, but it will bring more redundancy, because the logic has been implemented in `RexToLixTranslator.convert` already. 

Therefore, this PR handle it in a case-by-case way. I ever tried to unify these functions and combine them into one. But I do not know whether it is an appropriate way. Maybe @zabetak  , @julianhyde and @vlsi can provide some helpful advices.

danny0405 replied:
I don't think `RexToLixTranslator.convert`, `RexToLixTranslator.translateCast` and `Types.castIfNecessary` have any function overlaps, the first 2 are purely about the expression translate, they just produce the expression you want, but `Types.castIfNecessary` contains logic to decide if the is any necessity to do the cast, that is why i suggest you to move the logic.

If you think i'm wrong or other people can do it better, feel free to ignore this.

DonnyZone replied:
@danny0405 Thanks for discussion and suggestion, Danny. It is great helpful to me to make it more clear.

Current now, I'm a bit of confusing about the semantic differences between `cast `and `convert`. It seems in `Types.castIfNecessary` also conduct some converting work. 
For instance, from **_Integer_** to **_int_**, the cast will call `Expressions.unbox`. Meanwhile, we can also do it with `RexToLixTranslator.convert`. So I am curious about the design intention, but I can't find some documents.

danny0405 replied:
`CAST` and `CONVERT` are 2 different sql functions but have the same functionality.

DonnyZone replied:
Hi, @danny0405, I have updated the PR based on your recent changes.
There are some problems when move the following logic from `RexToLixTranslator.convert` into `Types.castIfNecessary`.
(1) It relies on SqlFunctions, which is impossible to call it in Linq4j.
(2) Make the logic duplicate.
Do you have some suggestions to handle it? Maybe we can unify them?
```
else if (toType == BigDecimal.class) {
      if (fromBox != null) {
        // E.g. from "Integer" to "BigDecimal".
        // Generate "x == null ? null : new BigDecimal(x.intValue())"
        return Expressions.condition(
            Expressions.equal(operand, RexImpTable.NULL_EXPR),
            RexImpTable.NULL_EXPR,
            Expressions.new_(
                BigDecimal.class,
                Expressions.unbox(operand, fromBox)));
      }
      if (fromPrimitive != null) {
        // E.g. from "int" to "BigDecimal".
        // Generate "new BigDecimal(x)"
        // Fix CALCITE-2325, we should decide null here for int type.
        return Expressions.condition(
            Expressions.equal(operand, RexImpTable.NULL_EXPR),
            RexImpTable.NULL_EXPR,
            Expressions.new_(
                BigDecimal.class,
                operand));
      }
      // E.g. from "Object" to "BigDecimal".
      // Generate "x == null ? null : SqlFunctions.toBigDecimal(x)"
      return Expressions.condition(
          Expressions.equal(operand, RexImpTable.NULL_EXPR),
          RexImpTable.NULL_EXPR,
          Expressions.call(
              SqlFunctions.class,
              "toBigDecimal",
              operand));
    }
```

danny0405 replied:
I kind of think we should invoke `RexToLixTranslator#convert` in the last step of the `Types#castIfNecessary` instead of the `Expressions.convert_`[1], what fo you think ?

[1] https://github.com/apache/calcite/blob/955d4ea7b30b85519529c1c77662fa04141bfa89/linq4j/src/main/java/org/apache/calcite/linq4j/tree/Types.java#L464

DonnyZone replied:
Yes, I agree. In my impression, there may be some tricks need to be handled. I will give a try. Thanks!

danny0405 commented on core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java:
```diff
@@ -546,6 +546,33 @@ public boolean isNullable() {
+    }
+
+    public int getScale() {
+      final SqlTypeName typeName =
```
Why add these 2 methods ? Where they are used ?

DonnyZone replied:
They are used in validation phase for **JavaType(BigDecimal)**.
For example:
```
select "primitiveByte" + "bigDecimal" as c
from "s"."everyTypes"
where "primitiveByte" <> 0
```
We will get the below stacktrace:

```
java.lang.AssertionError
	at org.apache.calcite.rel.type.RelDataTypeSystem.deriveDecimalPlusType(RelDataTypeSystem.java:151)
	at org.apache.calcite.sql.type.ReturnTypes.lambda$static$6(ReturnTypes.java:516)
	at org.apache.calcite.sql.type.SqlTypeTransformCascade.inferReturnType(SqlTypeTransformCascade.java:54)
	at org.apache.calcite.sql.type.SqlReturnTypeInferenceChain.inferReturnType(SqlReturnTypeInferenceChain.java:56)
	at org.apache.calcite.sql.SqlOperator.inferReturnType(SqlOperator.java:470)
	at org.apache.calcite.sql.SqlOperator.validateOperands(SqlOperator.java:437)
	at org.apache.calcite.sql.SqlOperator.deriveType(SqlOperator.java:516)
	at org.apache.calcite.sql.SqlBinaryOperator.deriveType(SqlBinaryOperator.java:144)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5590)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5577)
	at org.apache.calcite.sql.SqlCall.accept(SqlCall.java:139)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveTypeImpl(SqlValidatorImpl.java:1677)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveType(SqlValidatorImpl.java:1662)
	at org.apache.calcite.sql.SqlAsOperator.deriveType(SqlAsOperator.java:133)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5590)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$DeriveTypeVisitor.visit(SqlValidatorImpl.java:5577)
	at org.apache.calcite.sql.SqlCall.accept(SqlCall.java:139)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveTypeImpl(SqlValidatorImpl.java:1677)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.deriveType(SqlValidatorImpl.java:1662)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expandSelectItem(SqlValidatorImpl.java:465)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelectList(SqlValidatorImpl.java:4068)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3375)
```

DonnyZone replied:
As the comments in `ReflectiveSchemaTest` in the PR.
https://github.com/apache/calcite/pull/1375/files

danny0405 replied:
I still can not see any invocation from the stack trace you pasted. Maybe you can add the comments above the methods.

DonnyZone replied:
If there are some operands with Decimal types in RexCall, its return type is Decimal, and the validator will calculate the result's precision and scale during inferring return type.

However, as we can see in **everyTypes**, when the column is a reflected JavaType(BigDecimal), its `(precision, scale)=(PRECISION_NOT_SPECIFIED, SCALE_NOT_SPECIFIED) = (-1, Integer.MIN_VALUE)`.

During the calculation for the return type's precision and scale in `RelDataTypeSystem.deriveDecimalPlusType`, it leads to Assertions (e.g., `assert precision > 0`).

DonnyZone replied:
@danny0405 thanks for your suggestion, I will add some comments for this change.

danny0405 replied:
In CALCITE-2302, i get a default decimal type for the Java type and use it's precision/scale, i think you can do this also. [1]

[1] https://github.com/apache/calcite/pull/706/files#diff-eb76e2b906a6e4edbcc85056ad58ce75R143

DonnyZone replied:
Yes, we can do it in `RelDataTypeSystem`.  But we may also need to calculate JavaType's precision/scale in `deriveDecimalMultiplyType`,` deriveDecimalPlusType` and `deriveDecimalDivideType`.

DonnyZone commented on core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java:
```diff
@@ -503,7 +503,8 @@ private RelDataType decimalOf2(RelDataType type) {
     switch (typeName) {
     case DECIMAL:
-      return type;
+      return RelDataTypeFactoryImpl.isJavaType(type)
```
The type name of `JavaType(BigDecimal)` is also "**DECIMAL**", but its precision and scale are not specified, i.e., `(PRECISION_NOT_SPECIFIED, SCALE_NOT_SPECIFIED`)

danny0405 replied:
`createSqlType(SqlTypeName.DECIMAL)` returns a decimal with max precision and 0 as scale, like a `LONG` type, for java decimal, maybe we should return a max precision/scale decimal, use method `SqlTypeUtil#getMaxPrecisionScaleDecimal`.

DonnyZone replied:
Thanks! I will make the change.

danny0405 commented on core/src/test/java/org/apache/calcite/test/ReflectiveSchemaTest.java:
```diff
@@ -589,6 +592,14 @@ private void check(ResultSetMetaData metaData, String columnName,
+        CalciteAssert.that().withSchema("s", CATCHALL);
+    with.query("select \"wrapperDouble\" / \"bigDecimal\" as c\n"
+        + " from \"s\".\"everyTypes\"")
+        .runs();
```
We can also give some cases: 'bigdecimal / null', 'null/bigdecimal', 'bigdecimal/double'.

DonnyZone replied:
I find that `ReflectiveSchemaTest#testOp()` enumerates all type combinations for numeric operators ('+', '-', '*', '/'), including `null` content (e.g., `wrapperInteger`, `wrapperDouble`). I think it can cover these cases. So, maybe we can even remove this test case.
However, when I tested a `null` literal with the following test case.
```
  @Test public void testDivideDoubleBigDecimal() {
    final CalciteAssert.AssertThat with =
        CalciteAssert.that().withSchema("s", CATCHALL);
    with.query("select \"bigDecimal\" / null as c\n"
        + " from \"s\".\"everyTypes\"")
        .runs();
  }
```

It fails due to the **Assertion** below. When SqlBinaryOperator figuring out its `Monotonicity`, if the second operand is a Literal, it will cast the literal into BigDecimal. But the `Null` Literal can not be properly handled.
It seems to be the problem related with NULL literal, which demands a systematic solution.
Should we file a seperate Jira for this issue (it can also be trigger by 'wrapperLong'/null)?

```
java.lang.AssertionError: cannot cast null as class java.math.BigDecimal
	at org.apache.calcite.sql.SqlLiteral.getValueAs(SqlLiteral.java:351)
	at org.apache.calcite.sql.SqlCallBinding.getOperandLiteralValue(SqlCallBinding.java:217)
	at org.apache.calcite.sql.SqlBinaryOperator.getMonotonicity(SqlBinaryOperator.java:190)
	at org.apache.calcite.sql.SqlCall.getMonotonicity(SqlCall.java:182)
	at org.apache.calcite.sql.SqlCallBinding.getOperandMonotonicity(SqlCallBinding.java:188)
	at org.apache.calcite.sql.SqlAsOperator.getMonotonicity(SqlAsOperator.java:139)
	at org.apache.calcite.sql.SqlCall.getMonotonicity(SqlCall.java:182)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertSelectList(SqlToRelConverter.java:3961)
```

danny0405 replied:
Well, this exception is not expected, personally, i think the proper place to throw is:

1. The type checker: if we do not allow null literal operands for DIVIDE
2. The execution runtime, if we do not support a `number/null` pattern expression(we should do it if we already allows the pattern in the type checker)

Fire another issue is a more proper way to go, thanks.

DonnyZone commented:
@danny0405 Since we have implemented precision/scale for Java Type in CALCITE-2302(https://issues.apache.org/jira/browse/CALCITE-2302), and unified the type cast in [CALCITE-3414](https://issues.apache.org/jira/browse/CALCITE-3414), the issue can be fixed naturally without any blockers.

Therefore, for this JIRA, we only need to fix the problem for Java Decimal's precision and scale. 
I have rebased PR, could you take a look when you have time?
